larpandoracontent/LArTrackShowerId/TrackShowerIdFeatureTool.cc

```diff
@@ -199,147 +194,567 @@ void LinearFitFeatureTool::CalculateVariablesSlidingLinearFit(const pandora::Clu
-    m_nearbyClusterDistance(2.5f)
+ThreeDLinearFitFeatureTool::ThreeDLinearFitFeatureTool() :
+    m_slidingLinearFitWindow(3),
+    m_slidingLinearFitWindowLarge(10000)
 {
 }
 
 //------------------------------------------------------------------------------------------------------------------------------------------
 
-void NNearbyClustersFeatureTool::Run(SupportVectorMachine::DoubleVector &featureVector, const Algorithm *const pAlgorithm,
-    const pandora::Cluster *const pCluster)
+void ThreeDLinearFitFeatureTool::Run(SupportVectorMachine::DoubleVector &featureVector, const Algorithm *const pAlgorithm,
+const pandora::ParticleFlowObject *const pInputPfo)
 {
     if (PandoraContentApi::GetSettings(*pAlgorithm)->ShouldDisplayAlgorithmInfo())
         std::cout << "----> Running Algorithm Tool: " << this->GetInstanceName() << ", " << this->GetType() << std::endl;
 
-    ClusterList candidateClusters;
+    ClusterList clusterList;
+    LArPfoHelper::GetTwoDClusterList(pInputPfo, clusterList);
+    float diffWithStraightLineMean(0.f), maxFitGapLength(0.f), rmsSlidingLinearFit(0.f);
 
-    for (const std::string &clusterListName : m_clusterListNames)
+    for (const Cluster *const pCluster : clusterList)
     {
-        const pandora::ClusterList *pClusterList = nullptr;
-        PANDORA_THROW_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::GetList(*pAlgorithm, clusterListName, pClusterList));
+        float straightLineLengthLargeCluster(-1.f), diffWithStraightLineMeanCluster(-1.f), maxFitGapLengthCluster(-1.f), rmsSlidingLinearFitCluster(-1.f);
 
-        for (const Cluster *const pCandidateCluster : *pClusterList)
-        {
-            if ((LArClusterHelper::GetClusterHitType(pCluster)) != (LArClusterHelper::GetClusterHitType(pCandidateCluster)))
-                continue;
+        this->CalculateVariablesSlidingLinearFit(pCluster, straightLineLengthLargeCluster, diffWithStraightLineMeanCluster, maxFitGapLengthCluster, rmsSlidingLinearFitCluster);
 
-            if ((pCandidateCluster != pCluster) && (pCandidateCluster->GetNCaloHits() >= m_minClusterCaloHits))
-                candidateClusters.push_back(pCandidateCluster);
+        if (straightLineLengthLargeCluster > std::numeric_limits<double>::epsilon())
+        {
+            diffWithStraightLineMeanCluster  /= straightLineLengthLargeCluster;
+            maxFitGapLengthCluster           /= straightLineLengthLargeCluster;
+            rmsSlidingLinearFitCluster       /= straightLineLengthLargeCluster;
         }
-    }
 
-    int nNearbyClusters(0);
+        diffWithStraightLineMean   += diffWithStraightLineMeanCluster;
+        maxFitGapLength            += maxFitGapLengthCluster;
+        rmsSlidingLinearFit        += rmsSlidingLinearFitCluster;
+    }
 
-    for (const Cluster *const pCandidateCluster : candidateClusters)
+    const float nClusters(static_cast<float>(clusterList.size()));
+    if (nClusters > std::numeric_limits<double>::epsilon())
```

Comment:
Might be best to check that `clusterList` isn't empty (a concrete check). Inside `if` statement, can then cast `clusterList.size()` to a float and can then divide safely.